range_queries/persistent_segment_tree_with_lazy_propagation.cpp

```diff
@@ -0,0 +1,243 @@
+ * A persistent data structure always preserves the previous version of itself when it is modified.
+ * That is, a new version of the segment tree is generated after every update.
+ * It saves all previous versions of itself (before every update) to facilitate doing range queries in any version.
+ * More memory is used ,but the logarithmic time is preserved because the new version points to the same nodes, that the previous version points to, that are not affected by the update.
+ * That is, only the nodes that are affected by the update and their ancestors are copied. The rest is copied using lazy propagation in the next queries.
+ * Thus preserving the logarithmic time because the number of nodes copied after any update is logarithmic.
+ *
+ * @author [Magdy Sedra](https://github.com/MSedra)
+ */
+#include<iostream>
+#include<vector>
+#include <memory> /// to manage dynamic memory
+
+/// Range query here is range sum ,but the code can be modified to make different queries like range max or min
+class perSegTree {
+private:
+    class Node {
+    public:
+        std::shared_ptr<Node> left; /// pointer to the left node
+        std::shared_ptr<Node> right; /// pointer to the right node
+        int64_t val, prop; /// val is the value of the node (here equals to the sum of the leaf nodes children of that node), prop is the value to be propagated/added to all the leaf nodes children of that node
+
+        Node() {
+            val = prop = 0;
+            left = right = nullptr;
+        }
+    };
+
+    int n; /// number of elements/leaf nodes in the segment tree
+    std::vector<std::shared_ptr<Node>> ptrs; /// ptrs[i] holds a root pointer to the segment tree after the ith update. ptrs[0] holds a root pointer to the segment tree before any updates
+    std::vector<int> vec; /// values of the leaf nodes that the segment tree will be constructed with
+
+    /**
+     * @brief Creating a new node with the same values of curr node
+     * @param curr node that would be copied
+     * @returns the new node
+     */
+    std::shared_ptr<Node> newKid(std::shared_ptr<Node> const &curr) {
+        auto newNode = std::make_shared<Node>(Node());
+        newNode->left = curr->left;
+        newNode->right = curr->right;
+        newNode->prop = curr->prop;
+        newNode->val = curr->val;
+        return newNode;
+    }
+
+    /**
+     * @brief If there is some value to be propagated to the passed node, value is added to the node and the children of the node, if exist, are copied and the propagated value is also added to them
+     * @param i the left index of the range that the passed node holds its sum
+     * @param j the right index of the range that the passed node holds its sum
+     * @param curr pointer to the node to be propagated
+     * @returns void
+     */
+    void lazy(int i, int j, std::shared_ptr<Node> const &curr) {
```

Comment:
As these two variables are not being modified, consider making them a `const` and pass them by reference.
```suggestion
    void lazy(const int& i, const int& j, std::shared_ptr<Node> const &curr) {
```